Di changes

dazinator opened the pull request:
Reminder: Make sure to review CONTRIBUTING.MD and make sure you've sent in your signed contributor agreement.

Issue: https://github.com/MarimerLLC/csla/issues/1874

This PR:

- CSLA is no longer responsible for creating or disposing DI scopes.
- CSLA is given a "default" IServiceProvider - which it will use as a "fallback" if there is not "scope" specific IServiceProvider set in LocalContext. 
- The aspnet core version of ContextManager , has an additional fallback which saves the user having to set the IServiceProvider for the current request, the precedence goes:
    `var currentSp = LocalContext ?? HttpContext.RequestServices ?? _defaultSp;` - same concept for the System.Web version.

- Because CSLA doesn't ever create a scope, it doesn't ever need to dispose a scope either.

NOTE: I have been unable to run the tests on my local VS.. - all sorts therefore may be broken - but this hopefully is enough to discuss whether its worth polishing off.

- I throw an exception on @inject with a null service provider - but I have not added a resx error message.. I am not fluent in many languages..

dazinator commented:
Test failing due to the fact that we now throw when the IServiceProvider is null and try and @inject params. I'll fix the test to expect an exception now in this scenario.

dazinator commented:
@rockfordlhotka there are 8 tests failing on the appveyor build due to an obscure error. I am unable to run the tests locally at all (nothing happens in vs2019 pro test explorer window) and based on the failing test error in the app veyor log I am not sure what it could be - but it looks culture / resx related?

dazinator commented on Source/Csla.Shared/Reflection/ServiceProviderMethodCaller.cs:
```diff
@@ -403,14 +403,17 @@ public static async Task<object> CallMethodTryAsync(object obj, ServiceProviderM
 #if !NET40 && !NET45
-        var service = ApplicationContext.ScopedServiceProvider;        
+        var service = ApplicationContext.ScopedServiceProvider;   
+        if(service == null)
```
This qualifies as an unrelated change - let me know if you want it on another PR.

dazinator replied:
Sorry I missed this comment about throwing an nre.. yes I hadn't realise that would avoid the need for a translation, would have saved me a headache

rockfordlhotka commented:
@dazinator I think the culture/resource exception is masking the underlying issue, which is that the change broke method-level injection and is now throwing exceptions.

So there are two issues - one is clearly a missing resource that I'm sure has been a hidden bug for some time. The other is the underlying issue that the tests are failing.

You _should_ be able to run at least the netcore tests locally on your dev machine. Some of the netfx tests rely on LocalDb and other things being installed, but the netcore tests (iirc) are all pure in-memory.

rockfordlhotka commented:
I really don't think it will work to not provide a scoped provider btw. 

Without _some_ scoped provider I don't see how most data access scenarios are viable, because things like database connections are normally scoped.

In other words, though the current situation might be imperfect, there must be _some_ scoped provider available to any code running on the server-side data portal right?

dazinator commented:
The scoped provider needs to be set at the point where the scope is created. With my changes, csla will use this scoped provider if its set, if its not set it will fallback (either to checking for httpcontext, or the "default" application level service provider acquired on startup which is the root scope if you like).

The bit I havent looked at yet, is - for a remote data portal, there should be an incoming request and and outgoing response for that data portal operation. That whole operation csla should wrap in its own DI scope. I.e on a remote data portal server, when a request is received csla should create a new DI scope using the Default service provider, then set that as the current scoped service provider

rockfordlhotka commented:
> I throw an exception on @Inject with a null service provider - but I have not added a resx error message.. I am not fluent in many languages..

All that is required is to add the English text to the non-localized resx file. All other languages will get that English text until folks update the localized resx files with translations.

rockfordlhotka commented:
> The bit I havent looked at yet, is - for a remote data portal, there should be an incoming request and and outgoing response for that data portal operation. That whole operation csla should wrap in its own DI scope (similar to what aspnet core does for each request) I.e on a remote data portal server, when a request is received csla should create a new DI scope using the Default service provider, then set that as the current scoped service provider in LocalContext

I see what you are saying. This may be something that needs to be done in each _data portal host_, because I suspect the rules may vary by host.

In other words, if you are hosting in a console app or something, then it is up to you entirely. But if you are hosting in ASP.NET perhaps the runtime has already created a scope for you?

dazinator commented:
Ok I'll make that resx change. Im typing on a mobile and accidentally hit the close pr button and then my battery died whilst mid way through editing my last comment! The joys of tech

dazinator commented:
> I see what you are saying. This may be something that needs to be done in each data portal host, because I suspect the rules may vary by host.

> In other words, if you are hosting in a console app or something, then it is up to you entirely. But if you are hosting in ASP.NET perhaps the runtime has already created a scope for you?

Exactly

dazinator commented:
.. and the reason I used the fallback logic (if there is no scoped IServiceProvider set in local context) use the default IServiceProvider is.. in cases where csla is used in a simple console app or unit test, the user might not feel the need to create any new DI scope at all. Technically the root service provider should be perfectly valid for use in cases where csla is not being used inside any scope - if however there is some constraint in terms of dependencies not being able to be resolved from root scope these exceptions should be surfaced up to the user via DI as that's a constraint of DI and the particular service in question and not CSLA imho!

rockfordlhotka commented:
> > In other words, if you are hosting in a console app or something, then it is up to you entirely. But if you are hosting in ASP.NET perhaps the runtime has already created a scope for you?

> Exactly, as we know aspnet core does.

So my question remains: how do you get a reference to the scoped provider created by aspnet?

dazinator commented:
> > > In other words, if you are hosting in a console app or something, then it is up to you entirely. But if you are hosting in ASP.NET perhaps the runtime has already created a scope for you?
> 
> > Exactly, as we know aspnet core does.
> 
> So my question remains: how do you get a reference to the scoped provider created by aspnet?

Do you mean asp.net core? If so there are a couple of ways. If you mean legacy aspnet then I have no idea as I havent touched that in what feels like decades.

In asp.net core, I've adjusted the context manager so that if the user hasn't set a specific scope it will fall back to checking for an HttpContext (via IHttpContextAccessor) and then getting the scoped provider from HttpContext.RequestServices. if there is no active httpcontext it falls back to the default sp. This saves the user (or csla) from having to have middleware that sets the scoped sp at the start of every request (which is the second apprach). The second approach is as I posted on one of out recent discussions - you add middleware to the middleware pipeline that runs at the start of a request. The middleware gets the current httpcontext so again, it can set the csla current scoped sp from HttpContext.RequestServices.

dazinator commented:
Not sure if you support OWIN but if you do, you'd use the middleware approach to create a DI scope for each request. OWIN doesn't have the concept of an IServiceProvider that is managed by the platform / framework but if users are prepared to manage this themselves in there own owin based applications  (by creating a scope at start of request etc) they can use csla by giving csla the scoped provider that they create. Csla won't have any way to auto detect this.

rockfordlhotka commented:
I think the reason you can't run tests is related to something with the vs test runner. I figured this out long ago, changed the setting, and then promptly forgot what I did 😳

In the test runner you need to change the default for AnyCPU assemblies so they run in x64 mode

![image](https://user-images.githubusercontent.com/2333134/97499054-3b161580-193b-11eb-8a48-1e98ea4f74bf.png)


Then close and reopen the csla.test solution and the tests should run.

I'm not hitting the resource issue when running the tests interactively, so I'm not sure why the build server is having that issue. But the tests absolutely are failing, so something in the PR broken method invocation.

dazinator commented:
Ok thanks, I'll give that a go tomorrow!

dazinator commented:
I had some problems adding a new resx entry, after adding the entry to the .resx file in the netstandard project, and saving the change and rebuilding, the resx designer class wasn't being regenerated. I had to add this to the project file to get this to work:

```
 <ItemGroup>
    <Compile Update="Properties\Resources.Designer.cs">
      <DesignTime>True</DesignTime>
      <AutoGen>True</AutoGen>
      <DependentUpon>Resources.resx</DependentUpon>
    </Compile>
  </ItemGroup> 

  <ItemGroup>
    <EmbeddedResource Update="Properties\Resources.resx">
      <Generator>ResXFileCodeGenerator</Generator>
      <LastGenOutput>Resources.Designer.cs</LastGenOutput>
    </EmbeddedResource>
  </ItemGroup>

```

These entries also result in the designer file being moved in solution explorer to be nested under the resx file:

![image](https://user-images.githubusercontent.com/3176632/97560483-9f39e780-19d6-11eb-98ef-222d0cc98795.png)


I am using vs2019 pro.

dazinator commented:
@rockfordlhotka 
After the "Any CPU X64" change in test runner, I can run some of the tests - all of the "csla.test" ones are passing when running locally for me:

![image](https://user-images.githubusercontent.com/3176632/97564376-66047600-19dc-11eb-8e18-c3e76a254f69.png)

However all of those other tests, I am still unable to get running. The AppVeyor build is failing with what looks like a resource file problem, and I am not sure what that could be - ideally I think I need to get these other tests running locally. Any more ideas? Which version of VS do you use?

dazinator commented:
.. and here is some output from the test runner:

```

TestSettings will soon be deprecated for automated unit and functional testing scenarios. It is recommended that you use RunSettings. To learn more, see http://aka.ms/runsettings
Unable to load the test container 'C:\Users\DarrellTunnell\source\repos\csla\Bin\Debug\NET46\Csla.Data.EF5.dll' or one of its dependencies. Error details: System.IO.FileLoadException: Could not load file or assembly 'EntityFramework, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)

Unable to load the test container 'C:\Users\DarrellTunnell\source\repos\csla\Bin\Debug\NET46\Csla.Web.Mvc4.dll' or one of its dependencies. Error details: System.IO.FileLoadException: Could not load file or assembly 'System.Web.Mvc, Version=4.0.0.1, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)

Warning : A testsettings file or a vsmdi file is not supported with the MSTest V2 Adapter.
Test run will use DLL(s) built for framework .NETFramework,Version=v4.5 and platform X64. Following DLL(s) do not match framework/platform settings.
Csla.dll is built for Framework .NETFramework,Version=v4.0,Profile=Client and Platform AnyCPU.
Go to http://go.microsoft.com/fwlink/?LinkID=236877&clcid=0x409 for more details on managing these settings.

TestSettings will soon be deprecated for automated unit and functional testing scenarios. It is recommended that you use RunSettings. To learn more, see http://aka.ms/runsettings
Warning : A testsettings file or a vsmdi file is not supported with the MSTest V2 Adapter.
No test is available in C:\Users\DarrellTunnell\source\repos\csla\Bin\Debug\NET4\Csla.dll. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
Test run will use DLL(s) built for framework .NETFramework,Version=v4.5 and platform X64. Following DLL(s) do not match framework/platform settings.
Csla.Analyzers.Tests.dll is built for Framework .NETCoreApp,Version=v3.1 and Platform AnyCPU.
Go to http://go.microsoft.com/fwlink/?LinkID=236877&clcid=0x409 for more details on managing these settings.

TestSettings will soon be deprecated for automated unit and functional testing scenarios. It is recommended that you use RunSettings. To learn more, see http://aka.ms/runsettings
Warning : A testsettings file or a vsmdi file is not supported with the MSTest V2 Adapter.
No test is available in C:\Users\DarrellTunnell\source\repos\csla\Source\Csla.Analyzers\Csla.Analyzers.Tests\bin\Debug\netcoreapp3.1\Csla.Analyzers.Tests.dll. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.
TestSettings will soon be deprecated for automated unit and functional testing scenarios. It is recommended that you use RunSettings. To learn more, see http://aka.ms/runsettings
Warning : A testsettings file or a vsmdi file is not supported with the MSTest V2 Adapter.
No test is available in C:\Users\DarrellTunnell\source\repos\csla\Bin\Debug\NET45\Csla.dll. Make sure that test discoverer & executors are registered and platform & framework version settings are appropriate and try again.

```

rockfordlhotka commented:
You know, I think I had to delete the `*.testsettings` and `*.vsmdi` files from the root of the /Source directory.

I don't know why, because the tests run in my normal working directory - but when I cloned your fork they wouldn't run for me either. Changing the test runner to x86 and deleting those files got them all running for me.

dazinator commented:
> You know, I think I had to delete the `*.testsettings` and `*.vsmdi` files from the root of the /Source directory.
> 
> I don't know why, because the tests run in my normal working directory - but when I cloned your fork they wouldn't run for me either. Changing the test runner to x86 and deleting those files got them all running for me.

Thanks - deleting those files, and setting test runner back to "Any Cpu" worked!

dazinator commented:
@rockfordlhotka 
As of latest push, I am left with two tests that fail locally for EF 4, but it looks like they are looking for a particular file called "testConnectionStrings" to exist on my machine which doesn't:


>  Unable to open configSource file 'testConnectionStrings.config'. (C:\Users\DarrellTunnell\source\repos\csla\Source\TestResults\DarrellTunnell_DESKTOP-UH729T5 2020-10-30 11_54_01\Out\Csla.Data.EF4.Test.DLL.config line 7).

I think this is a seperate issue unrelated to my changes!

dazinator commented:
Also, for this AppVeyor build failure:

> A total of 1 test files matched the specified pattern.
  ! DashboardFailureCounter
  ! DashboardRecentActivity
  X NoServiceProvider_Throws [114ms]
  Error Message:
   Assert.ThrowsException failed. Threw exception MissingManifestResourceException, but exception InjectException was expected. 
Exception Message: Could not find any resources appropriate for the specified culture or the neutral culture.  Make sure "Csla.Properties.Resources.resources" was correctly embedded or linked into assembly "Csla" at compile time, or that all the satellite assemblies required are loadable and fully signed.
Stack Trace:    at System.Resources.ManifestBasedResourceGroveler.HandleResourceStreamMissing(String fileName)
   at System.Resources.ManifestBasedResourceGroveler.GrovelForResourceSet(CultureInfo culture, Dictionary`2 localResourceSets, Boolean tryParents, Boolean createIfNotExists, StackCrawlMark& stackMark)
   at System.Resources.ResourceManager.InternalGetResourceSet(CultureInfo requestedCulture, Boolean createIfNotExists, Boolean tryParents, StackCrawlMark& stackMark)
   at System.Resources.ResourceManager.InternalGetResourceSet(CultureInfo culture, Boolean createIfNotExists, Boolean tryParents)
   at System.Resources.ResourceManager.GetString(String name, CultureInfo culture)
   at Csla.Reflection.ServiceProviderMethodCaller.CallMethodTryAsync(Object obj, ServiceProviderMethodInfo method, Object[] parameters) in C:\projects\csla\Source\Csla.Shared\Reflection\ServiceProviderMethodCaller.cs:line 415
   at Csla.Test.DataPortal.ServiceProviderInvocationTests.<>c__DisplayClass1_0.<<NoServiceProvider_Throws>b__0>d.MoveNext() in C:\projects\csla\Source\csla.netcore.test\DataPortal\ServiceProviderInvocationTests.cs:line 44
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.VisualStudio.TestTools.UnitTesting.Assert.ThrowsExceptionAsync[T](Func`1 action, String message, Object[] parameters)
  Stack Trace:
     at Csla.Test.DataPortal.ServiceProviderInvocationTests.NoServiceProvider_Throws() in C:\projects\csla\Source\csla.netcore.test\DataPortal\ServiceProviderInvocationTests.cs:line 44

I am not sure why the appveyor test has a problem with locating the resx resource so I am looking into this.
I am guessing its something to do with the build and the way resx files are shuffled around. The test passes locally though. I originally added the following to the netstandard project so that after editing the resx file, code gen would give me an updated designer class:

```
<ItemGroup>
    <Compile Update="Properties\Resources.Designer.cs">
      <DesignTime>True</DesignTime>
      <AutoGen>True</AutoGen>
      <DependentUpon>Resources.resx</DependentUpon>
    </Compile>
  </ItemGroup> 

  <ItemGroup>
    <EmbeddedResource Update="Properties\Resources.resx">
      <Generator>PublicResXFileCodeGenerator</Generator>
      <LastGenOutput>Resources.Designer.cs</LastGenOutput>
    </EmbeddedResource>
  </ItemGroup>

```

But looking more at the project file I think the workflow you were after is to make changes in the Csla.Shared.Resources resx files, and then have those copied into the netstandard project:

```
 <Target Name="PrecompileScript" BeforeTargets="BeforeBuild">
    <Exec Command="copyResources.cmd" />
    <Exec Command="del Properties\CslaPrivate.resx" />
  </Target>

```

So I'll try reverting my project file changes, and using that approach. Still - as the test passes locally its not yet clear why appveyor would be different.

dazinator commented:
Hmm - sorry, I am out of ideas on the appveyor build front!

1. Tests all pass locally for me. I did have an odd issue where the "testConnections" file was just not being output with the build output, and this caused large amounts of tests to suddenly fail - despite it being set to CopyLocal=Always. Restarting VS appeard to fix this :-(
 
![image](https://user-images.githubusercontent.com/3176632/97705738-784de500-1aac-11eb-9eef-e696bdb68c17.png)

2. I've "Undone" any project file changes I made to do with resx but the appveyor issue still exists. I understand the tests might be running under a different culture, but as my entry is in the Resources.resx file I would expect this to work still, so something about the way the resx's are being linked / compiled under the appveyor build may be the issue?

dazinator commented:
I'm tempted to remove the change I made to throw this exception resulting in the need for the resx entry - just to see if the build then passes - that change is somewhat unrelated anyway.

dazinator commented:
@rockfordlhotka 
I've removed the logic to throw an exception on @inject with a null service provider from this PR and reverted it to what it was previously, because attempting to add a new resx entry was breaking the appveyor build for unknown reasons - and was kind of unrelated to the main focus of this PR.

rockfordlhotka reviewed (changes requested):
I think that new inject exception type should be removed as well?

dazinator commented:
> I think that new inject exception type should be removed as well?

Done

ashishsinha24 commented:
I am using the 5.4.1 Preview & .Net 5.0

I am getting this strange error.

Could not find the resource "Csla.Properties.Resources.resources" among the resources "" embedded in the assembly "Csla", nor among the resources in any satellite assemblies for the specified culture. Perhaps the resources were embedded with an incorrect name.
